src/cp/cp.rs

```diff
@@ -658,27 +686,94 @@ fn copy(sources: &[Source], target: &Target, options: &Options) -> CopyResult<()
-
     for source in sources {
         if seen_sources.contains(source) {
             show_warning!("source '{}' specified more than once", source.display());
 
-        } else if let Err(error) = copy_source(source, target, &target_type, options) {
-            show_error!("{}", error);
-            match error {
-                Error::Skipped(_) => (),
-                _ => non_fatal_errors = true,
+        } else {
+            let mut found_hard_link = false;
+            if preserve_hard_links {
+                #[cfg(unix)]
+                unsafe {
+                    let dest = construct_dest_path(source, target, &target_type, options)?;
+                    let mut stat = mem::zeroed();
+                    let src_path = CString::new(Path::new(&source.clone()).as_os_str().to_str().unwrap()).unwrap();
+
+                    if libc::lstat(src_path.as_ptr(), &mut stat) < 0 {
+                        return Err(format!("cannot stat {:?}: {}", src_path, std::io::Error::last_os_error()).into());
+                    }
+
+                    let inode = stat.st_ino;
+                    for hard_link in &hard_links {
+                        if hard_link.1 == inode {
+                            std::fs::hard_link(hard_link.0.clone(), dest.clone());
+                            found_hard_link = true;
+                        }
+                    }
+                    if stat.st_nlink > 1 && !found_hard_link {
+                        hard_links.push((dest.clone().to_str().unwrap().to_string(), inode));
+                    }
+                }
+                #[cfg(windows)]
+                {
+                    unsafe {
+                        if !source.is_dir() {
+                            let dest = construct_dest_path(source, target, &target_type, options)?;
+                            let handle = CreateFile2(CString::new(Path::new(&source.clone()).as_os_str().to_str().unwrap()).unwrap().as_ptr() as *const u16,
+                                                                    winapi::winnt::GENERIC_READ,
+                                                                    winapi::winnt::FILE_SHARE_READ,
+                                                                    0,
+                                                                    std::ptr::null_mut());
+                            let file_info = std::mem::uninitialized();
+                            if GetFileInformationByHandle(handle, file_info) != 0 {
+                                return Err(format!("cannot get file information {:?}: {}", source, std::io::Error::last_os_error()).into());
+                            }
+
+                            let file_index = ((*file_info).nFileIndexHigh, (*file_info).nFileIndexLow);
+                            for hard_link in &hard_links {
+                                if (hard_link.1).0 == file_index.0 && (hard_link.1).1 == file_index.1 {
```

Comment:
Are the parentheses here necessary?

Reply:
If you mean the parentheses for the tuples then yes, they are necessary